[FSDP] Improve the documentation of state_dict

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73532
* __->__ #73453
* #73366

Add more documentation to state_dict and local_state_dict including
basic examplesAdd more documentation to state_dict and local_state_dict
including basic examples.

Differential Revision: [D34483812](https://our.internmc.facebook.com/intern/diff/D34483812/)

rohan-varma committed:
[FSDP] Improve the documentation of state_dict

Add more documentation to state_dict and local_state_dict including
basic examplesAdd more documentation to state_dict and local_state_dict
including basic examples.

Differential Revision: [D34483812](https://our.internmc.facebook.com/intern/diff/D34483812/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -854,8 +854,36 @@ def _post_state_dict_hook(
+        The entry point of all three FSDP state_dict APIs. By default, calling
+        `state_dict` on an FSDP module will result in FSDP attempting to bring
+        the entire (nested) model into memory and taking the local model's
+        `state_dict` on every rank, which could result in OOM if the model
```
does this needs to be double quote to make it render like code?

mrshenli commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -854,8 +854,36 @@ def _post_state_dict_hook(
+        >>> from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+        >>> from torch.distributed.fsdp import StateDictType
+        >>> torch.cuda.set_device(device_id)
+        >> my_module = nn.Linear(...)
```
missing one >

mrshenli commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -976,8 +1004,40 @@ def load_state_dict(
+        >>> from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+        >>> from torch.distributed.fsdp import StateDictType
+        >>> torch.cuda.set_device(device_id)
+        >> my_module = nn.Linear(...)
```
ditto

mrshenli commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -976,8 +1004,40 @@ def load_state_dict(
+        >>> full_dict.keys()
+        >>> odict_keys(['weight', 'bias'])
+        >>> # using local state dict
+        >> local_state_dict = checkpoint['local_state_dict]
```
ditto

rohan-varma committed:
Update on "[FSDP] Improve the documentation of state_dict"


Add more documentation to state_dict and local_state_dict including
basic examplesAdd more documentation to state_dict and local_state_dict
including basic examples.

Differential Revision: [D34483812](https://our.internmc.facebook.com/intern/diff/D34483812/)

[ghstack-poisoned]
